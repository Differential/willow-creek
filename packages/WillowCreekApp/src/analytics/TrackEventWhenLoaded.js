import { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { track } from './index';

export default class TrackEventWhenLoaded extends PureComponent {
  componentDidMount() {
    if (this.props.loaded) {
      this.track();
    }
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.loaded && this.props.loaded) {
      this.track();
    }
  }

  get trackClient() {
    return this.props.track || track;
  }

  track() {
    const { eventName, properties } = this.props;
    return this.trackClient({ eventName, properties });
  }

  render() {
    return null;
  }
}

TrackEventWhenLoaded.propTypes = {
  loaded: PropTypes.bool.isRequired,
  eventName: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  properties: PropTypes.any,
  track: PropTypes.func,
};
