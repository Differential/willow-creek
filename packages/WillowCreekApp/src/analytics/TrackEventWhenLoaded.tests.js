import React from 'react';
import renderer from 'react-test-renderer';

import TrackEventWhenLoaded from './TrackEventWhenLoaded';

const trackMock = jest.fn();

describe('The TrackEventWhenLoaded component', () => {
  afterEach(() => {
    trackMock.mockClear();
  });
  it('should track an event if mounted with loaded=true', () => {
    renderer.create(
      <TrackEventWhenLoaded
        loaded
        eventName="Track Something"
        properties={{ ThingId: '1234' }}
        track={trackMock}
      />
    );
    expect(trackMock.mock.calls).toMatchSnapshot();
  });

  it('should track an event if mounted with loaded=false and then switching to true', () => {
    const tree = renderer.create(
      <TrackEventWhenLoaded
        loaded={false}
        eventName="Track Something"
        properties={{ ThingId: '1234' }}
        track={trackMock}
        key="tracker"
      />
    );
    expect(trackMock.mock.calls).toMatchSnapshot();
    tree.update(
      <TrackEventWhenLoaded
        loaded
        eventName="Track Something"
        properties={{ ThingId: '1234' }}
        track={trackMock}
        key="tracker"
      />
    );
    expect(trackMock.mock.calls).toMatchSnapshot();
  });
});
