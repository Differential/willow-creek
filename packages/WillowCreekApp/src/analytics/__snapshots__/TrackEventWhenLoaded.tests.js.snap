// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The TrackEventWhenLoaded component should track an event if mounted with loaded=false and then switching to true 1`] = `Array []`;

exports[`The TrackEventWhenLoaded component should track an event if mounted with loaded=false and then switching to true 2`] = `
Array [
  Array [
    Object {
      "eventName": "Track Something",
      "properties": Object {
        "ThingId": "1234",
      },
    },
  ],
]
`;

exports[`The TrackEventWhenLoaded component should track an event if mounted with loaded=true 1`] = `
Array [
  Array [
    Object {
      "eventName": "Track Something",
      "properties": Object {
        "ThingId": "1234",
      },
    },
  ],
]
`;
